### PR TITLE
Add property filters to sessions

### PR DESCRIPTION
**Pull request by timgl**
## Changes

*Please describe.*  
*If this affects the front-end, include screenshots.*  

## Checklist

- [ ] All querysets/queries filter by Team (if this PR affects any querysets/queries)
- [ ] Backend tests (if this PR affects the backend)
- [ ] Cypress E2E tests (if this PR affects the front and/or backend)

### CODE REVIEW

File: frontend/src/scenes/sessions/sessionsTableLogic.ts [hunk 1]
```diff
@@ -73,37 +94,41 @@ export const sessionsTableLogic = kea<sessionsTableLogicType<Moment, SessionType
...
     urlToAction: ({ actions, values }) => ({
-        '/sessions': (_: any, { date }: { date: string }) => {
+        '/sessions': (_: any, { date, properties }: { date: string; properties: Record<string, any> }) => {
             const newDate = date ? moment(date).startOf('day') : moment().startOf('day')
-            if (!values.selectedDate || values.selectedDate.format('YYYY-MM-DD') !== newDate.format('YYYY-MM-DD')) {
-                actions.dateChanged(newDate)
+            if (
+                !values.selectedDate ||
+                values.selectedDate.format('YYYY-MM-DD') !== newDate.format('YYYY-MM-DD') ||
+                JSON.stringify(properties || {}) !== JSON.stringify(values.filters)
```
Comment (macobo) [inline frontend/src/scenes/sessions/sessionsTableLogic.ts:119]: There already exists an `objectsEqual` method in frontend/src/lib/utils.js, perhaps we should use that?

Comment (macobo) [inline frontend/src/scenes/sessions/sessionsTableLogic.ts:119]: Q: JSON.stringfy(x) is dependent  on key order, this means:

```
JSON.stringify({ a: 1, b: 2 }) !== JSON.stringify({ b: 2, a: 1 })
```

Is that relevant in this case?

File: frontend/src/scenes/sessions/sessionsTableLogic.ts [hunk 2]
```diff
@@ -35,9 +50,9 @@ export const sessionsTableLogic = kea<sessionsTableLogicType<Moment, SessionType
         setNextOffset: (nextOffset: number | null) => ({ nextOffset }),
         fetchNextSessions: true,
         appendNewSessions: (sessions) => ({ sessions }),
-        dateChanged: (date: Moment | null) => ({ date }),
         previousDay: true,
         nextDay: true,
+        setFilters: (filters: Record<string, any>, selectedDate: Moment | null) => ({ filters, selectedDate }),
```
Comment (macobo) [inline frontend/src/scenes/sessions/sessionsTableLogic.ts:55]: Usage of `any` is generally discouraged within ts, but I guess to type this properly we'd need to port PropertyFilters.js and some other code as well as well?

File: frontend/src/scenes/sessions/sessionsTableLogic.ts [hunk 3]
```diff
@@ -73,37 +94,41 @@ export const sessionsTableLogic = kea<sessionsTableLogicType<Moment, SessionType
             }
             actions.appendNewSessions(response.result)
         },
-        dateChanged: () => {
-            actions.loadSessions(true)
+        setFilters: () => {
             actions.setNextOffset(null)
+            actions.loadSessions(true)
```
Comment (macobo) [inline frontend/src/scenes/sessions/sessionsTableLogic.ts:99]: Is the action order change significant here?

File: frontend/src/scenes/sessions/sessionsTableLogic.ts (hunk 2 above)
Comment (mariusandra) [reply frontend/src/scenes/sessions/sessionsTableLogic.ts:55]: Would definitely be cool to have a completely typed filter! Indeed lots of "legacy" code to upgrade to make this work well. Until then `any` is probably fine here. There's also `unknown`, but that's a small difference.

**Comment by mariusandra** [comment]
@timgl we worked together on this at the offsite to try to avoid the situation where the request in `loadSessions` is called twice or even three times due to various things updating the URL, state, etc. 

I now came upon an extremely simple solution that would solve all problems without any of the hacks we implemented:

```ts
export const sessionsTableLogic = kea<sessionsTableLogicType<Moment, SessionType>>({
    loaders: ({ actions, values, props }) => ({
        sessions: {
            __default: [] as SessionType[],
            loadSessions: async (_: any, breakpoint) => {
                const { selectedDateURLparam } = values
                const params = toParams({
                    date_from: selectedDateURLparam,
                    date_to: selectedDateURLparam,
                    offset: 0,
                    distinct_id: props.personIds ? props.personIds[0] : '',
                    properties: values.filters,
                })
                // add this!
                await breakpoint(10)

                const response = await api.get(`api/insight/session/?${params}`)
                breakpoint()
                if (response.offset) {
                    actions.setNextOffset(response.offset)
                }
                return response.result
            },
        },
    }) 
)}
```

Adding a simple breakpoint before the `api.get` debounces the query and if another `loadSesssions` is called within 10ms since the first one, no call is made. This is a simple and cool debouncing pattern that's super easy to implement with kea loaders. The number could even be smaller, since all the actions fire right after each other.

Not saying it's needed to redo this PR, but something to definitely keep in mind for the future :).

**Comment by timgl** [comment]
@macobo @mariusandra Thanks both. Marius, I've implemented your suggestion which made things a lot easier. I'm now also using PropertyFilter's built in urlToAction which means back and forward history navigation works too!

**Comment by mariusandra** [comment]
Great! One comment (nit, feel free to ignore) coming back to what Karl said. It might be a good idea to replace all those filter types `Record<string, any>`s with some `PropertyFilterType` from `src/types`, even if for now it's just:

```ts
export type PropertyFilterType = Record<string, any>
```

It can be converted into a real type some time later. This will however already help future refactors.

**Comment by timgl** [comment]
Fixed! Also cleaned up filters/properties confusion

**Review by mariusandra — APPROVED** [review]
LGTM

**Review by macobo — APPROVED** [review]

**Pull request merged**